Add trait for filtering pull requests based on label

cwholmes opened the pull request:
# Description

This change introduces a filter (include/exclude) trait for pull requests by label. This can be useful for projects that want to limit builds of pull request to only the pull requests containing a specific label or to exclude pull requests (such as WIP pull requests).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Automated tests have been added to exercise the changes
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

# Documentation changes
- [ ] Link to jenkins.io PR, or an explanation for why no doc changes are needed

# Users/aliases to notify

cwholmes commented:
@bitwiseman Thanks for letting me know `Head` changed logic. I have moved the labels retrieval into the filter itself rather than storing them on the `Head` object. I wasn't sure on how to test that a change in labels wouldn't result in the `Head` being considered changed, but the change I made should mitigate that need.

bitwiseman commented:
@cwholmes  I assume you've tested locally?

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/WildcardPullRequestLabelFilterTrait/help-includes.html:
```diff
@@ -0,0 +1,29 @@
+ ~ THE SOFTWARE.
+ -->
+<div>
+    Space-separated list of project pull request label patterns to consider.
+    You may use <code>*</code> as a wildcard; for example: <code>jenkins *-plugin</code>
+ 
+    If the pull request has no labels it will only be included if the includes filter is <code>*</code>.
```
```suggestion
    Space-separated list of project pull request label patterns to consider.
    You may use <code>*</code> as a wildcard; for example: <code>jenkins *-plugin</code>
    <br />
    If the pull request has any label that matches any include pattern, it will be included.
    <br />
    If the pull request has no labels it will only be included if the includes filter is <code>*</code>.
```

cwholmes commented:
> @cwholmes I assume you've tested locally?

I have tested the changes with a local jenkins deployment.

The trait in the list of traits:

<img width="263" alt="trait-in-list" src="https://user-images.githubusercontent.com/14796526/120904151-3a034000-c610-11eb-8c45-2630501d2d86.png">

The default trait being added:

<img width="263" alt="default-trait-added" src="https://user-images.githubusercontent.com/14796526/120904154-412a4e00-c610-11eb-99ab-3b8355e0e1a3.png">

Updated with an include only:

<img width="288" alt="include-added" src="https://user-images.githubusercontent.com/14796526/120904170-599a6880-c610-11eb-8d3f-e96931087213.png">

PRs not being included:

<img width="481" alt="prs-not-included" src="https://user-images.githubusercontent.com/14796526/120904180-63bc6700-c610-11eb-975d-924cd795b593.png">

Updated with an exclusion only:

<img width="236" alt="exclusion-added" src="https://user-images.githubusercontent.com/14796526/120904183-6c14a200-c610-11eb-96a6-6921991f0bf8.png">

PRs being excluded:

<img width="401" alt="prs-excluded" src="https://user-images.githubusercontent.com/14796526/120904188-7767cd80-c610-11eb-95d1-741e60f91727.png">

Updated with both include and exclude:

<img width="243" alt="include-and-exclude-added" src="https://user-images.githubusercontent.com/14796526/120904190-83538f80-c610-11eb-9517-e51dab070e7e.png">

PRs being included and excluded:

<img width="497" alt="prs-included-and-excluded" src="https://user-images.githubusercontent.com/14796526/120904217-b5fd8800-c610-11eb-9a3e-52f64f9395fb.png">

KalleOlaviNiemitalo commented on src/main/java/org/jenkinsci/plugins/github_branch_source/WildcardPullRequestLabelFilterTrait.java:
```diff
@@ -0,0 +1,168 @@
+  @Symbol({"labelPullRequestFilter"})
+  @Extension
+  @Selection
+  public static class DescriptorImpl extends SCMSourceTraitDescriptor {
+    public DescriptorImpl() {}
+
+    public String getDisplayName() {
+      return Messages.WildcardPullRequestLabelFilterTrait_DisplayName();
+    }
+  }
```
Please make this descriptor only applicable to GitHub branch sources, see [JENKINS-66503](https://issues.jenkins.io/browse/JENKINS-66503).

cwholmes replied:
This has been addressed and the ignore drafts trait has been fixed here https://github.com/jenkinsci/github-branch-source-plugin/pull/458

jtnord commented:
I think this is obselete by already provided functionality of the [Github Additional Traits plugin](https://github.com/jenkinsci/github-additional-traits-plugin)?

jtnord reviewed (commented):
I think this is obsolete by already provided functionality of the [Github Additional Traits plugin](https://github.com/jenkinsci/github-additional-traits-plugin)?

simonlsk commented:
@jtnord , this plugin is not available on Jenkins (in the manage plugins screen), I'm currently using version 2.332. 
In general, I couldn't find any trait for Pull Requests that prevent triggering a build that makes sense (PR title, Description content, etc...). I can filter by the branch name but it is "PR-777" which is not useful. 
I currently add logic inside my jenkinfile pipeline, but Jenkins still queues the PR, a dedicated instance connects, the pipeline starts, etc... this is a significant overhead for me. 
Am I missing something?

jtnord commented:
> @jtnord , this plugin is not available on Jenkins (in the manage plugins screen), I'm currently using version 2.332. In general, I couldn't find any trait for Pull Requests that prevent triggering a build that makes sense (PR title, Description content, etc...). I can filter by the branch name but it is "PR-777" which is not useful. I currently add logic inside my jenkinfile pipeline, but Jenkins still queues the PR, a dedicated instance connects, the pipeline starts, etc... this is a significant overhead for me. Am I missing something?

it seems that plugin was never released :-( 
however there is https://github.com/jenkinsci/github-label-filter-plugin that maybe fits the bill?

simonlsk commented:
> however there is https://github.com/jenkinsci/github-label-filter-plugin that maybe fits the bill?

Yes, this is useful, thank you!

jtnord commented:
closing based on recent comments.  Please re-open with comments if I have miss understood.
